### ISSUE TITLE
no-shadow if a type alias and function argument share the same name

### ISSUE THREAD

**Issue by SebastianNiemann** [eslint/typescript-eslint-parser#459]
**What version of TypeScript are you using?**
2.7.1

**What version of `typescript-eslint-parser` are you using?**
14.0.0

**What code were you trying to parse?**

```ts
type foo = any;
function bar(foo: any) {}
```

**What did you expect to happen?**
Not a `no-shadow` error.

**What happened?**

```
2:14  error  'foo' is already declared in the upper scope  no-shadow
```

In contrast, this works as expected (no such error):

```ts
type foo = any;
let foo: any;
```

**Comment by JamesHenry** [comment]
I think a custom rule may need to be created in the [plugin](https://github.com/nzakas/eslint-plugin-typescript) to handle this. It would be fantastic if one of the people affected by this issue could contribute to it.

**Comment by SebastianNiemann** [comment]
Will I would like to help you out on this, I am currently not familiar enough with this project's code base and too busy to find enough time to provide you with a working pull request soon.

I think it would take me at least four weeks, before I could even start look into it.

However, maybe you could clear something up for me. If I got it correctly, the current behaviour is invalid and therefore caused by a bug at either this side or https://github.com/nzakas/eslint-plugin-typescript. Following this up, I did not understood how a custom rule will help. I would assume that some of the existing code needs to be fixed, correct? Do you have a hunch about which of these two projects could be the cause for this? Or in in other words, where I should contribute to?

**Comment by SebastianNiemann** [comment]
Thanks 👍 Coincidentally, I just looked into https://github.com/nzakas/eslint-plugin-typescript/blob/master/lib/rules/no-unused-vars.js and assumed it works this way. So this makes perfectly sense to me.

Could you also give me some quick information about how the plugin and core rules interact with each other? Are the plugin's rules replacing the core's ones, or are both called? And if the later one is true, in which ordering?

From what I saw under https://github.com/nzakas/eslint-plugin-typescript/blob/master/lib/rules/no-unused-vars.js, it seems to be no replacement for https://github.com/eslint/eslint/blob/master/lib/rules/no-unused-vars.js, and only fixes the TS specific cases. I would also expect that it is called before the core rules, so that the line within the plugin

```js
variables[i].eslintUsed = true;
```

is called before `collectUnusedVariables` executes in the core rule, correct?

Following this train of thought, it seems I need to write a rule, that changes something used by `checkForShadows`, to get around the error message.

If I reduce the function to its important parts, it boils down to me as

```js
function checkForShadows(scope) {
  const variables = scope.variables;

  for (let i = 0; i < variables.length; ++i) {
    const variable = variables[i];

    /* .. */

    // Gets shadowed variable.
    const shadowed = astUtils.getVariableByName(scope.upper, variable.name);

    if (shadowed /* ... */) {
      context.report({/* ... */});
    }
  }
}
```

So it seems that `astUtils.getVariableByName` should return no result for `type foo = any;` (from my example at the start of this issue) and if I look into this within https://github.com/eslint/eslint/blob/master/lib/ast-utils.js, it seems that `if (variable) {` and more specific `scope.set.get(name)` needs to evaluate to `false`.

However, now I am lost how on to do this, without removing the type from the scope or rewriting its representation, avoiding to fail other rules which rely on this information.
